File: web/DBManager.php
<?
class DeviceType {
	const PowerSwitch = 0;
	const Sensor = 1;
}

class DeviceState {
	const Off = 0;
	const On = 1;
}

class NodeType {
	const PublicIP = 0;
	const Master = 1;
	const Slave = 2;
}

# Retrieving local ipaddress of PiHome, assuming the same as the webserver ipaddress
define('SCRIPT_IP', $_SERVER['SERVER_ADDR']);
define('SCRIPT_PORT', 50000);

# Function to connect to MySQL database on localhost with predefined credentials
function db_connect() {
	$link = mysql_connect('localhost', 'ceng499', 'ceng499');
	if(!$link) {
		die('Could not connect: 50000' . mysql_error());
	}
	
	$db_selected = mysql_select_db('pihome', $link);
	if(!$db_selected) {
		die ('Can\'t use pihome: ' . mysql_error());
	}
}

# Function to execute MySQL query
function db_query($query) {
	$result = mysql_query($query);
	
	if(!$result) {
		$message  = 'Invalid query: ' . mysql_error() . "\n";
		$message .= 'Whole query: ' . $query;
		die($message);
	}
	
	return $result;
}

function getUsers() {
	return db_query("SELECT username as Username, is_admin as 'Administrator' FROM Users");
}

function changePass($username, $pass) {
	$query = sprintf("UPDATE Users SET pass = SHA1('%s') WHERE username ='%s'",
		mysql_real_escape_string($pass),
		mysql_real_escape_string($username)
	);
	
	db_query($query);
}

function isAdmin() {
	if(!isset($_SERVER['PHP_AUTH_USER'])) return 0;
	
	$query = sprintf("SELECT is_admin FROM Users WHERE username ='%s'",
		mysql_real_escape_string($_SERVER['PHP_AUTH_USER'])
	);
	
	$result = db_query($query);
	$row = mysql_fetch_row($result);
	return $row[0];
}

###################################### Camera node related functions
# Function to get a list of all nodes
function getNodes() {
	return db_query("SELECT nodename,ipaddress FROM Nodes");
}

# Function to get all node names from Nodes
function getCamNamesResult() {
	return db_query("SELECT nodename FROM Nodes");
}

# Function to get ipaddress of a node
function getCamIP($name) {
	$query = sprintf("SELECT ipaddress FROM Nodes WHERE nodename ='%s'",
		mysql_real_escape_string($name));
	$result = db_query($query);
	$row = mysql_fetch_row($result);
	return $row[0];
}

# Function to add new node to database
function addNode($nodename, $nodeaddress) {
	$query = sprintf("INSERT INTO Nodes VALUES('%s','%s')",
		mysql_real_escape_string($nodename),
		mysql_real_escape_string($nodeaddress)
	);
	
	db_query($query);
}

# Function to change camera node name
function changeNodeName($dname, $dnewname) {
	# Change Node name in Nodes table
	$query = sprintf("UPDATE Nodes SET nodename='%s' WHERE nodename='%s'",
		mysql_real_escape_string($dnewname),
		mysql_real_escape_string($dname)
	);
	
	db_query($query);

	# Change Node name in Playbacks table
	$query = sprintf("UPDATE Playbacks SET nodename='%s' WHERE nodename='%s'",
		mysql_real_escape_string($dnewname),
		mysql_real_escape_string($dname)
	);
	
	db_query($query);
}

# Function to remove node
function removeNode($nodename, $nodeaddress) {
	# Remove Node from Nodes table
	$query = sprintf("DELETE FROM Nodes WHERE nodename='%s' AND ipaddress='%s'",
		mysql_real_escape_string($nodename),
		mysql_real_escape_string($nodeaddress)
	);
	
	db_query($query);
}

################################### Trigger related functions
# Function to add new door trigger to database
function addDoorTrigger($dserial) {
	$query = sprintf("INSERT INTO DoorTriggers VALUES(%s, 0, 0)",
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to add new switch trigger to database
function addSwitchTrigger($dserial) {
	$query = sprintf("INSERT INTO SwitchTriggers VALUES(%s, 0, 0)",
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to return list of door triggers
function getDoorTriggers() {
	return db_query("SELECT lpad(hex(doorserial),16,'0') as doorserial, lpad(hex(switchserial),16,'0') as switchserial, action FROM DoorTriggers");
}

# Function to return list of switch triggers
function getSwitchTriggers() {
	return db_query("SELECT lpad(hex(switchserial),16,'0') as switchserial, ondailytime, offdailytime FROM SwitchTriggers");
}

# Function to remove door trigger
function removeDoorTrigger($dserial) {
	$query = sprintf("DELETE FROM DoorTriggers WHERE doorserial=%s",
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to remove switch trigger
function removeSwitchTrigger($dserial) {
	$query = sprintf("DELETE FROM SwitchTriggers WHERE switchserial=%s",
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to update door switch
function updateDoorSwitch($dserial, $sserial) {
	$query = sprintf("UPDATE DoorTriggers SET switchserial=%s WHERE doorserial=%s",
		mysql_real_escape_string($sserial),
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to update door action
function updateDoorAction($dserial, $action) {
	$query = sprintf("UPDATE DoorTriggers SET action=%s WHERE doorserial=%s",
		mysql_real_escape_string($action),
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to update switch on time daily
function updateOnTime($dserial, $ondailytime) {
	$query = sprintf("UPDATE SwitchTriggers SET ondailytime=%s WHERE switchserial=%s",
		mysql_real_escape_string($ondailytime),
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to update switch off time daily
function updateOffTime($dserial, $offdailytime) {
	$query = sprintf("UPDATE SwitchTriggers SET offdailytime=%s WHERE switchserial=%s",
		mysql_real_escape_string($offdailytime),
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

################################### Device related functions
# Function to get list of devices
function getDevicesResult() {
	return db_query("SELECT lpad(hex(serial),16,'0') as Serial, type as Type, name as Name, status as Status, message as Message, active as Active FROM Devices");
}

# Function to add new device to database
function addDevice($dserial, $dname) {
	$query = sprintf("INSERT INTO Devices VALUES(%s, 5, '%s', 0, 0,'New')",
		mysql_real_escape_string($dserial),
		mysql_real_escape_string($dname)
	);
	
	db_query($query);
}

# Function to change device name
function changeDeviceName($dserial, $dname) {
	# Change device name in Devices table
	$query = sprintf("UPDATE Devices SET name='%s' WHERE serial=%s",
		mysql_real_escape_string($dname),
		mysql_real_escape_string($dserial)
	);
	
	db_query($query);
}

# Function to remove device
function removeDevice($dserial, $dname) {
	$query = sprintf("DELETE FROM Devices WHERE serial=%s AND name='%s'",
		mysql_real_escape_string($dserial),
		mysql_real_escape_string($dname)
	);
	
	db_query($query);
}

# Function to update device active state
function toggleDeviceActive($dserial, $dname, $dactive) {
	$query = sprintf("UPDATE Devices SET active=%s WHERE serial=%s AND name='%s'",
		mysql_real_escape_string($dactive),
		mysql_real_escape_string($dserial),
		mysql_real_escape_string($dname)
	);
	
	db_query($query);
}

############################ Record time related functions
# Function to get record time
function getRecordTime() {
	return db_query("SELECT recordtime FROM RecordTime");
}

# Function to change record time
function updateRecordTime($newtime, $oldtime) {
	$query = sprintf("UPDATE RecordTime SET recordtime=%s WHERE recordtime=%s",
		mysql_real_escape_string($newtime),
		mysql_real_escape_string($oldtime)
	);
	
	db_query($query);
}

############################ Email related functions
# Function to get all emails
function getEmails() {
	return db_query("SELECT email FROM Emails");
}

# Function to add new email
function addEmail($email) {
	$query = sprintf("INSERT INTO Emails VALUES('%s')",
		mysql_real_escape_string($email)
	);
	
	db_query($query);
}

# Function to remove email
function removeEmail($email) {
	$query = sprintf("DELETE FROM Emails WHERE email='%s'",
		mysql_real_escape_string($email)
	);
	
	db_query($query);
}

# Function to change email
function changeEmail($newemail, $oldemail) {
	$query = sprintf("UPDATE Emails SET email='%s' WHERE email='%s'",
		mysql_real_escape_string($newemail),
		mysql_real_escape_string($oldemail)
	);
	
	db_query($query);
}


############################################## Playback related functions
# Function to return list of nodename and playback folders
function getCamPlaybacksResult() {
	return db_query("SELECT nodename, recordfolder FROM Playbacks");
}

############### Function to send command to remore Pi via socket
function sendPiCam($cam, $msg) {
	$ip = getCamIP($cam);
	if($ip == '') die("cam not found: " . $cam);
	
	$sock = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
	if(!$sock) die('socket_create');
	
	if(!socket_connect($sock, $ip, 44444)) {
		socket_close($sock);
		die('socket_connect: ' . $ip . ', ' . $msg);
	}
	
	socket_send($sock, $msg, strlen($msg), 0);
	
	socket_close($sock);
}

# Function to send STARTPLAYBACK command
function startPlayback($cam, $path) {
	sendPiCam($cam, 'STARTPLAYBACK,' . $path);
}

# Function to send INIT command
function camReset($cam) {
	sendPiCam($cam, 'INIT');
}

# Function to send DELPLAYBACK
function deletePlayback($cam, $path) {
	$query = sprintf("DELETE FROM Playbacks WHERE nodename='%s' AND recordfolder='%s'",
		mysql_real_escape_string($cam),
		mysql_real_escape_string($path)
	);
	db_query($query);
	
	sendPiCam($cam, 'DELPLAYBACK,' . $path);	
}

# Function to send STARTRECORD command
function startRecord($cam, $path) {
	sendPiCam($cam, 'STARTRECORD,' . $path);
	
	$query = sprintf("INSERT INTO Playbacks VALUES('%s', '%s')",
		mysql_real_escape_string($cam),
		mysql_real_escape_string($path)
	);
	
	db_query($query);
}


################################################# Function to send command to central PiHome
function sendCommandToPiHome($command, $param) {
	# Check for non-empty command and parameter
	if(strlen($command) == 0 or strlen($param) == 0) die('Command and Parameter are empty' . $command . $param);
	
	# Build command message consists of "command param"
	$message = $command. '/pi/' . $param;
	
	# Create one-time socket to local PiHome.py to send command to it
	$sock = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
	if(!$sock) die('Cannot create socket');
	
	if(!socket_connect($sock, SCRIPT_IP, SCRIPT_PORT)) {
		socket_close($sock);
		die('Cannot connect to PiHome.py ip and port');
	}
	
	# Send command through socket
	socket_send($sock, $message, strlen($message), 0);
	
	# Do not wait for socket response, close right away.
	# This may need to change if we want to wait for confirmation from PiHome.py
	socket_close($sock);
}

# Function to set switch state through PiHome, should use sendCommandToPiHome function
function setDeviceState($dserial, $state) {
	#$non_num_pattern = "/[^0-9A-Z]/";
	
	#if(preg_match($non_num_pattern, $dserial) || preg_match($non_num_pattern, $state)) {
	#	die('Non decimal character detected');
	#}
	
	$state = intval($state, 10);
	
	if($state == DeviceState::Off) $msg = 'off';
	else if($state == DeviceState::On) $msg = 'on';
	else die('Invalid state' . $state);
	
	$msg .= '/pi/' . $dserial;
	
	$sock = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
	if(!$sock) die('1');
	
	if(!socket_connect($sock, SCRIPT_IP, SCRIPT_PORT)) {
		socket_close($sock);
		die('2');
	}
	
	socket_send($sock, $msg, strlen($msg), 0);
	
	socket_close($sock);
}

db_connect();

?>

File: scripts/DBManager.py
#!/usr/bin/python
"""
Python script for managing database
"""
import MySQLdb


# Function to create MySQL database connection
def getDBCon():
	return MySQLdb.connect(host="localhost", # your host, usually localhost
		                   user="ceng499", # your username
		                   passwd="ceng499", # your password
		                   db="pihome") # name of the data base
		                        
# Function to create fresh database tables
def initDatabase():
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Create Devices table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE 'Devices'")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE Devices")
	dbcur.execute("CREATE TABLE Devices(serial BIGINT UNSIGNED PRIMARY KEY, type TINYINT, name VARCHAR(20), status TINYINT, active TINYINT, message VARCHAR(50))")

	# Create Credential table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE 'Credential'")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE Credential")
	dbcur.execute("CREATE TABLE Credential(username VARCHAR(20) PRIMARY KEY, password VARCHAR(20))")
	dbcur.execute("INSERT INTO Credential VALUES('pihome','pihomepass')")

	# Create Emails table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE 'Emails'")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE Emails")
	dbcur.execute("CREATE TABLE Emails(email VARCHAR(20) PRIMARY KEY)")
	dbcur.execute("INSERT INTO Emails VALUES('pimation@email.com')")

	# Create RecordTime table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE 'RecordTime'")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE RecordTime")
	dbcur.execute("CREATE TABLE RecordTime(recordtime TINYINT PRIMARY KEY)")
	dbcur.execute("INSERT INTO RecordTime VALUES(10)")
	
	# Create Nodes table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE 'Nodes'")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE Nodes")
	dbcur.execute("CREATE TABLE Nodes(nodename VARCHAR(10) PRIMARY KEY, ipaddress VARCHAR(15))")

	# Create Playbacks table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE 'Playbacks'")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE Playbacks")
	dbcur.execute("CREATE TABLE Playbacks(nodename VARCHAR(10), recordfolder VARCHAR(24))")

	# Create DoorTriggers table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE DoorTriggers")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE DoorTriggers")
	dbcur.execute("CREATE TABLE DoorTriggers(doorserial BIGINT UNSIGNED PRIMARY KEY, switchserial BIGINT UNSIGNED, action TINYINT)")

	# Create SwitchTriggers table, drop previous table if existed
	dbcur.execute("SHOW TABLES LIKE SwitchTriggers")
	result=dbcur.fetchone()
	if not (result is None or result[0] is None):
		dbcur.execute("DROP TABLE SwitchTriggers")
	dbcur.execute("CREATE TABLE SwitchTriggers(switchserial BIGINT UNSIGNED PRIMARY KEY, ondailytime TINYINT, offdailytime TINYINT)")
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to add new device
#db.addDevice(0013a20040a57ae9,0,'Switch',0,1,'New')
def addDevice(_serial, _type, _name, _status, _active, _message):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Check if device is already existed
	dbcur.execute("SELECT * FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		# Insert new entry to Devices table
		dbcur.execute("INSERT INTO Devices VALUES(%s,%s,%s,%s,%s,%s)", (_serial,_type,_name,_status,_active,_message))
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to remove device
def removeDevice(_serial):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Remove device if existed
	dbcur.execute("SELECT * FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Device %s does not exist" % _serial)
	else:		
		dbcur.execute("DELETE FROM Devices WHERE serial=%s", _serial)
		
	# Commit, close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()
	
# Function to get device information
def getDevice(_serial):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Get device
	dbcur.execute("SELECT serial,type,name,status,active,message FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()

	# Close db cursor, connection
	dbcur.close()
	dbcon.close()

	# Return result	
	if result is None or result[0] is None:
		print("Device %s does not exist" % _serial)
		return None
	else:		
		# Return device info
		return result


# Function to get all devices
def getDevices():
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Find all devices
	dbcur.execute("SELECT * FROM Devices")
	results=dbcur.fetchall()
		
	# Close db cursor, connection
	dbcur.close()
	dbcon.close()
	return results
		
# Function to update device status
def updateDeviceStatus(_serial, _status):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Check if device exists
	dbcur.execute("SELECT * FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Device %s does not exist" % _serial)
	else:		
		# Update status of device
		dbcur.execute("UPDATE Devices SET status=%s WHERE serial=%s", (_status,_serial))
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to update device message
def updateDeviceMessage(_serial, _message):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Check if device exists
	dbcur.execute("SELECT * FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Device %s does not exist" % _serial)
	else:		
		# Update status of device
		dbcur.execute("UPDATE Devices SET message=%s WHERE serial=%s", (_message,_serial))
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to activate or deactivate device
def updateDeviceActive(_serial, _active):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Check if device exists
	dbcur.execute("SELECT * FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Device %s does not exist" % _serial)
	else:		
		# Update active of device
		dbcur.execute("UPDATE Devices SET active=%s WHERE serial=%s", (_active,_serial))
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()
	
# Function to change device name
def changeDeviceName(_serial, _name):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Check if device exists
	dbcur.execute("SELECT * FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Device %s does not exist" % _serial)
	else:		
		# Change name of device
		dbcur.execute("UPDATE Devices SET name=%s WHERE serial=%s", (_name,_serial))
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to change device type
def changeDeviceType(_serial, _type):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Check if device exists
	dbcur.execute("SELECT * FROM Devices WHERE serial=%s", _serial)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Device %s does not exist" % _serial)
	else:		
		# Change type of device
		dbcur.execute("UPDATE Devices SET type=%s WHERE serial=%s", (_type,_serial))
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()
	
# Function to change password
def changePassword(_user, _oldpass, _newpass):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Check if user credential is correct
	dbcur.execute("SELECT * FROM Credential WHERE username=%s AND password=%s", (_user,_oldpass))
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Incorrect username and password %s %s" % (_user,_oldpass))
	else:		
		# Update new password
		dbcur.execute("UPDATE Credential SET password=%s WHERE username=%s", (_newpass,_user))
		
	# Commit querry and close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to get all emails
def getEmails():
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Find all emails
	dbcur.execute("SELECT * FROM Emails")
	results=dbcur.fetchall()

	# Close db cursor, connection
	dbcur.close()
	dbcon.close()
	
	return results

# Function to get record time
def getRecordTime():
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Find all emails
	dbcur.execute("SELECT recordtime FROM RecordTime")
	results=dbcur.fetchall()

	# Close db cursor, connection
	dbcur.close()
	dbcon.close()
	
	return results
	
# Function to add email
def addEmail(_email):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Add new email if not existed
	dbcur.execute("SELECT email FROM Emails WHERE email=%s", _email)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		dbcur.execute("INSERT INTO Emails VALUES(%s)", _email)
	# Close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to remove email
def removeEmail(_email):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Remove email if existed
	dbcur.execute("SELECT email FROM Emails WHERE email=%s", _email)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Email %s does not exist" % _email)
	else:
		dbcur.execute("DELETE FROM Emails WHERE email=%s", _email)
	# Close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to get all nodes
def getNodes():
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Find all emails
	dbcur.execute("SELECT * FROM Nodes")
	results=dbcur.fetchall()

	# Close db cursor, connection
	dbcur.close()
	dbcon.close()
	
	return results


# Function to get node information
def getNode(_nodename):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Get node info if existed
	dbcur.execute("SELECT * FROM Nodes WHERE nodename=%s", _nodename)
	result=dbcur.fetchone()

	# Close db cursor, connection
	dbcur.close()
	dbcon.close()
	
	if result is None or result[0] is None:
		print("Node %s does not exist" % _nodename)
		return None
	else:
		return result

			
# Function to add new node
def addNode(_nodename, _ipadd):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Add new node if not existed
	dbcur.execute("SELECT * FROM Nodes WHERE nodename=%s", _nodename)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		dbcur.execute("INSERT INTO Nodes VALUES(%s,%s)", (_nodename,_ipadd))

	# Commit, close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to remove node
def removeNode(_nodename):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Remove node if existed
	dbcur.execute("SELECT * FROM Nodes WHERE nodename=%s", _nodename)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Node %s does not exist" % _nodename)
	else:
		dbcur.execute("DELETE FROM Nodes WHERE nodename=%s", _nodename)

	# Commit, close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to update node ipaddress and port
def updateNode(_nodename, _ipadd):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Update node if existed
	dbcur.execute("SELECT * FROM Nodes WHERE nodename=%s", _nodename)
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Node %s does not exist" % _nodename)
	else:
		dbcur.execute("UPDATE Nodes SET ipaddress=%s WHERE nodename=%s", (_ipadd,_nodename))

	# Commit, close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()

# Function to get all playbacks
def getPlaybacks():
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Find all emails
	dbcur.execute("SELECT * FROM Playbacks")
	results=dbcur.fetchall()

	# Close db cursor, connection
	dbcur.close()
	dbcon.close()
	
	return results
		
# Function to add new playback
def addPlayback(_nodename, _recordfolder):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Add new playback if not existed
	dbcur.execute("SELECT * FROM Playbacks WHERE nodename=%s AND recordfolder=%s", (_nodename,_recordfolder))
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		dbcur.execute("INSERT INTO Playbacks VALUES(%s,%s)", (_nodename,_recordfolder))

	# Commit, close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()	
	
# Function to remove playback
def removePlayback(_nodename,_recordfolder):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Remove playback if existed
	dbcur.execute("SELECT * FROM Playbacks WHERE nodename=%s AND recordfolder=%s", (_nodename,_recordfolder))
	result=dbcur.fetchone()
	if result is None or result[0] is None:
		print("Playback %s,%s does not exist" % (_nodename,_recordfolder))
	else:
		dbcur.execute("DELETE FROM Playbacks WHERE nodename=%s AND recordfolder=%s", (_nodename,_recordfolder))

	# Commit, close db cursor, connection
	dbcon.commit()
	dbcur.close()
	dbcon.close()
	
# Function to get door trigger information
def getDoorTrigger(_serial):
	# Get database connection and cursor
	dbcon = getDBCon()
	dbcur = dbcon.cursor()
	
	# Get device
	dbcur.execute("SELECT doorserial,switchserial,action FROM DoorTriggers WHERE doorserial=%s", _serial)
	result=dbcur.fetchone()

	# Close db cursor, connection
	dbcur.close()
	dbcon.close()

	# Return result	
	if result is None or result[0] is None:
		print("DoorTrigger %s does not exist" % _serial)
		return None
	else:		
		# Return device info
		return result
		
